### PR TITLE
Include recreated system databases during selective restore.

### PR DESCRIPTION
**Pull request by pgstef**
Upon issue #986, the system databases `template0`, `template1` and `postgres` should be included by default when using `--db-include`.

Let's then retreive their id in the manifest and include them.

As pointed by Stephen, users might want to force exclude some databases (like `postgres`), so adding a `--db-exclude` option to do that. This new option allows to exclude any database, even the "system" ones. 

Tests added and should run fine:
```
$ pgbackrest/test/test.pl --vm-build --vm=f32
$ pgbackrest/test/test.pl --vm-out --dev --module=command --test=restore --coverage-only --vm=f32
...
2021-02-22 16:29:35.819 P00   INFO: P1-T1/1 - vm=f32, module=command, test=restore (10.81s)
2021-02-22 16:29:36.729 P00   INFO: tested modules have full coverage
2021-02-22 16:29:36.730 P00   INFO: writing C coverage report
2021-02-22 16:29:36.999 P00   INFO: TESTS COMPLETED SUCCESSFULLY (13s)
```

### CODE REVIEW

File: src/command/restore/restore.c [hunk 1]
```diff
@@ -1248,9 +1296,13 @@ restoreSelectiveExpression(Manifest *manifest)
...
                 const String *db = strLstGet(dbList, dbIdx);
 
-                // Only user created databases can be zeroed, never system databases
-                if (cvtZToUInt64(strZ(db)) >= PG_USER_OBJECT_MIN_ID)
+                // Include system databases by default
+                if (strLstExists(excludeDbIdList, db) ||
+                	(strLstSize(includeDbIdList) > 0 && cvtZToUInt64(strZ(db)) >= PG_USER_OBJECT_MIN_ID && 
+                	!strLstExists(includeDbIdList, db) && !strLstExists(systemDbIdList, db)))
                 {
+                    LOG_DETAIL_FMT("database with id %s will be zeroed", strZ(db));
```
Comment (woblerr) [inline src/command/restore/restore.c:1304]: Is it possible to print database name in addition to database id? Since this is done for system databases: [L1230](https://github.com/pgbackrest/pgbackrest/pull/1307/files#diff-9cfdda13c7fe664728b8fb1c44176c0408e96ac403d1e15e2f96265e5ddfa35fR1230)
The output will become more useful.

Comment (pgstef) [reply src/command/restore/restore.c]: that's possible yes, it requires to loop through the manifest-db-list. See commit 3e12e9c that adds this info.

File: src/command/restore/restore.c [hunk 2]
```diff
@@ -1248,9 +1300,20 @@ restoreSelectiveExpression(Manifest *manifest)
             {
                 const String *db = strLstGet(dbList, dbIdx);
 
-                // Only user created databases can be zeroed, never system databases
-                if (cvtZToUInt64(strZ(db)) >= PG_USER_OBJECT_MIN_ID)
+                // Include system databases by default
+                if (strLstExists(excludeDbIdList, db) ||
+                	(strLstSize(includeDbIdList) > 0 && cvtZToUInt64(strZ(db)) >= PG_USER_OBJECT_MIN_ID && 
+                	!strLstExists(includeDbIdList, db) && !strLstExists(systemDbIdList, db)))
```
Comment (cmwshang) [inline src/command/restore/restore.c:1306]: I think we have more String Lists than we need. The end result is that we want dbList to contain the IDs of the databases that we will zero. The problem with the original code is that it kept the system databases in the dbList and then thought they could just be filtered out by checking their ID. That, of course, is the bug. So our end goal should be that the dbList must only contain the IDs of the databases to zero. So first we should check  that the user does not provide the same database in the --db-include and --db-exclude list (as you did - good job!). Then any --db-exclude databases must **remain** in the dbList. Any system database that is **not** in the exclude list, must be **removed** from the dbList. Any --db-include database must be **removed** from the list. Then the entire IF condition goes away here because all the databases in the dbList are going to be zeroed.

**Review by cmwshang — CHANGES_REQUESTED** [review]

File: src/command/restore/restore.c [hunk 3]
```diff
@@ -1213,13 +1213,65 @@ restoreSelectiveExpression(Manifest *manifest)
...
+                if (systemDb != NULL)
+                {
+                    // LOG_DETAIL_FMT("system database id to include : %u (%s)", systemDb->id, strZ(systemDbName));
+                    strLstAdd(systemDbIdList, varStrForce(VARUINT(systemDb->id)));
+                }
+            }
+
+            strLstSort(systemDbIdList, sortOrderAsc);
+            if(strLstSize(systemDbIdList) > 0)
+                LOG_DETAIL_FMT("system databases to include: %s", strZ(strLstJoin(systemDbIdList, ", ")));
```
[the remark this answers is not in the document]
Comment (pgstef) [reply src/command/restore/restore.c:1237]: Hm, you're right indeed. I guess the `database ID (NAME) will be zeroed` log will be enough.

File: src/command/restore/restore.c [hunk 4]
```diff
@@ -1213,13 +1213,65 @@ restoreSelectiveExpression(Manifest *manifest)
...
 
-            // Remove included databases from the list
+            // Get system db ids in case users would have manually re-created it
+            StringList *systemDbList = strLstNew();
+            strLstAdd(systemDbList, STRDEF("template0"));
+            strLstAdd(systemDbList, STRDEF("template1"));
+            strLstAdd(systemDbList, STRDEF("postgres"));
+            StringList *systemDbIdList = strLstNew();
+
+            for (unsigned int systemDbListIdx = 0; systemDbListIdx < strLstSize(systemDbList); systemDbListIdx++)
```
[the remark this answers is not in the document]
Comment (pgstef) [reply src/command/restore/restore.c:1223]: I found it easier to read by building the name list, especially if we wanted to easily add more db names later (or even add option to configure it) but that shouldn't happen often indeed.
I changed that. Thanks for the suggestion.

File: src/command/restore/restore.c (hunk 2 above)
Comment (pgstef) [reply src/command/restore/restore.c]: Hm, is it a memory/perf issue to have an extra small string list to make the code more readable ?

Since we have to keep track of the databases to exclude (via dblist minus include-list or via exclude-list), I didn't managed to remove more than the `includeDbIdList`. 
See commit f494c9a about it.

I hope this is approaching what you had in mind :-)

File: src/command/restore/restore.c [hunk 5]
```diff
@@ -1248,9 +1298,19 @@ restoreSelectiveExpression(Manifest *manifest)
...
+                    (strLstSize(includeList) > 0 && cvtZToUInt64(strZ(db)) >= PG_USER_OBJECT_MIN_ID))
                 {
+                    // Get db name to LOG
+                    for (unsigned int dbIdx = 0; dbIdx < manifestDbTotal(manifest); dbIdx++)
+                    {
+                        const ManifestDb *dbInfo = manifestDb(manifest, dbIdx);
+
+                        if (strEq(db, varStrForce(VARUINT(dbInfo->id))))
+                            LOG_DETAIL_FMT("database %s (%s) will be zeroed", strZ(db), strZ(dbInfo->name));
+                    }
```
Comment (cmwshang) [inline src/command/restore/restore.c:1312]: I'm a little worried about this FOR loop. If we have a million databases, we're just doing this to get the name to put into a detail-level log message. I know someone thought it might be nice, but I don't think we should take the hit. Instead, I think before the loop, we should output a `LOG_DETAIL_FMT("databases to be zeroed: %s", strZ(strLstJoin(dbList, ", ")));`
We always output the command and options provided to console if they have at least INFO set and always to the log file and they will (or should be) running the `pgbackrest info` command with the --set option on the backup they will use to restore if they are going to do selective restore and they can get the OIDs from that.

File: src/command/restore/restore.c [hunk 6]
```diff
@@ -1232,11 +1279,14 @@ restoreSelectiveExpression(Manifest *manifest)
...
+                // Error if the db id is in the exclude list
+                if(strLstExists(excludeDbIdList, includeDb))
+                    THROW_FMT(DbInvalidError, "database to include '%s' is in the exclude list", strZ(includeDb));
+
                 // Error if the db is a system db
-                if (cvtZToUInt64(strZ(includeDb)) < PG_USER_OBJECT_MIN_ID)
+                if (cvtZToUInt64(strZ(includeDb)) < PG_USER_OBJECT_MIN_ID || strLstExists(systemDbIdList, includeDb))
                     THROW(DbInvalidError, "system databases (template0, postgres, etc.) are included by default");
 
-                // Remove from list of DBs to zero
```
Comment (cmwshang) [inline src/command/restore/restore.c:1239]: Can we add back this comment? The comment at the beginning of the loop is great, but also keeping this causes less churn in the diff's when reviewing - I had to ask myself "Wait, why was this removed?"

**Review by cmwshang — CHANGES_REQUESTED** [review]
We are almost there - I can't tell you how wonderful it is to have you contributing - your instincts are really good, Stefan!

File: src/command/restore/restore.c [hunk 7]
```diff
@@ -1248,9 +1298,19 @@ restoreSelectiveExpression(Manifest *manifest)
             {
                 const String *db = strLstGet(dbList, dbIdx);
 
-                // Only user created databases can be zeroed, never system databases
-                if (cvtZToUInt64(strZ(db)) >= PG_USER_OBJECT_MIN_ID)
+                // Include system databases by default
+                if (strLstExists(excludeDbIdList, db) ||
+                    (strLstSize(includeList) > 0 && cvtZToUInt64(strZ(db)) >= PG_USER_OBJECT_MIN_ID))
```
[the remark this answers is not in the document]
Comment (pgstef) [reply src/command/restore/restore.c:1303]: Not if we only use `--db-exlude`. I struggled a lot to figure this out but we won't remove the "to keep" databases from dbList if we don't use `--db-include`.
I'll simplify the IF regarding the system dbs but we'll have to keep it in case includeList is empty.

File: src/command/restore/restore.c (hunk 6 above)
Comment (pgstef) [reply src/command/restore/restore.c:1239]: Ok

File: src/command/restore/restore.c (hunk 5 above)
Comment (pgstef) [reply src/command/restore/restore.c:1312]: I'm also in favor of using ids everywhere in the log details, for consistency reasons.
But as mentioned above, dbList would contain databases that don't have to be zeroed in case we use `--db-exclude` alone. So your log detail would probably not be accurate. But I'll remove the loop and the name ;-)

File: src/command/restore/restore.c [hunk 8]
```diff
@@ -1213,15 +1213,62 @@ restoreSelectiveExpression(Manifest *manifest)
...
+            unsigned int systemDbFound = 0;
+            unsigned int systemDbIdx = 0;
+            StringList *systemDbIdList = strLstNew();
+
+            while (systemDbIdx < manifestDbTotal(manifest) && systemDbFound < 3)
+            {
+                const ManifestDb *systemDb = manifestDb(manifest, systemDbIdx);
+                if (strEqZ(systemDb->name, "template0") ||
+                    strEqZ(systemDb->name, "template1") ||
+                    strEqZ(systemDb->name, "postgres"))
```
[the remark this answers is not in the document]
Comment (pgstef) [reply src/command/restore/restore.c:1250]: Agreed. I'll add it.
The contributing guide says the CR should be after the 1st condition. And having "template1" + "postgres" on the same line went over 132, so CR too. But I'll adjust to minimize CR's then :-)

File: src/command/restore/restore.c [hunk 9]
```diff
@@ -1232,11 +1279,14 @@ restoreSelectiveExpression(Manifest *manifest)
                     includeDb = varStrForce(VARUINT(db->id));
```
[the remark this answers is not in the document]
Comment (pgstef) [reply src/command/restore/restore.c:1279]: ok

File: src/command/restore/restore.c [hunk 10]
```diff
@@ -1232,11 +1279,14 @@ restoreSelectiveExpression(Manifest *manifest)
...
                 }
 
+                // Error if the db id is in the exclude list
+                if(strLstExists(excludeDbIdList, includeDb))
+                    THROW_FMT(DbInvalidError, "database to include '%s' is in the exclude list", strZ(includeDb));
+
                 // Error if the db is a system db
-                if (cvtZToUInt64(strZ(includeDb)) < PG_USER_OBJECT_MIN_ID)
+                if (cvtZToUInt64(strZ(includeDb)) < PG_USER_OBJECT_MIN_ID || strLstExists(systemDbIdList, includeDb))
```
[the remark this answers is not in the document]
Comment (pgstef) [reply src/command/restore/restore.c:1287]: yaeh, adding all the system dbs to the list simplifies a lot of conditions :-)

File: src/command/restore/restore.c (hunk 7 above)
[the remark this answers is not in the document]
Comment (pgstef) [reply src/command/restore/restore.c:1303]: Saying it all loud made me realize I could remove the IF by adding:
```
// Only exclude specified db in case no db to include has been provided
if (strLstSize(includeList) <= 0)
    dbList = strLstDup(excludeDbIdList);
```

Fixing the tests and I'll push the modification ;-)

File: src/command/restore/restore.c (hunk 10 above)
[the remark this answers is not in the document]
Comment (pgstef) [reply src/command/restore/restore.c]: Actually, this one is pretty tricky. 
It breaks the MockAllTest because we don't generate the `[db]` section in the manifest.
I tried to add (line 1077 - "error on system id"):
```
$oManifest{&MANIFEST_SECTION_BACKUP_DB}{"template1"}{&MANIFEST_KEY_DB_ID} = 1;
$oManifest{&MANIFEST_SECTION_BACKUP_DB}{"template1"}{&MANIFEST_KEY_DB_LAST_SYSTEM_ID} = 12168;
```
But no luck. I'm not comfortable enough with the perl tests at this point. Would you have any idea to fix this test?

Many thanks in advance :-)

[the remark this answers is not in the document]
Comment (cmwshang) [reply src/command/restore/restore.c]: No worries. I should have some time tomorrow to take a look.

[the remark this answers is not in the document]
Comment (cmwshang) [reply src/command/restore/restore.c]: I pulled and began working on the branch in my own repository (https://github.com/cmwshang/pgbackrest/tree/pgstef-issues-986) so you can compare. I haven't yet completed the tests but will let you know when I do. Thanks again.

**Comment by cmwshang** [comment]
After further discussions, we decided we should split this PR and get the original issue solved first. The db-exclude option adds a lot of code and tests for something that we feel we need to think more about. So, @pgstef if you could pull out the db-exclude code and leave in the code and tests for getting the systemIds from the names, that would be great.

**Comment by woblerr** [comment]
@cmwshang  If I understand correctly, first of all we need to add possible database excluding by using the 'db-exclude' 
flag, including system databases, for example, 'postgres'. Further studies of changing the logic of pgbacrest and untying it from the number '16384' will be performed outside the scope of this PR. Will it be a separate issue or something similar? Could you confirm this?

**Comment by cmwshang** [comment]
> @cmwshang If I understand correctly, first of all we need to add possible database excluding by using the 'db-exclude' flag, including system databases, for example, 'postgres'. Further studies of changing the logic of pgbacrest and untying it from the number '16384' will be performed outside the scope of this PR. Will it be a separate issue or something similar? Could you confirm this?

Not quite. The selective restore feature is being enhanced to ensure that all databases will be excluded (zeroed) that are not provided with a specified --db-include option, with the exception of system databases (postgres, template0, template1 and any with a system OID). We will be fixing the issue we currently have that if a system database (say postgres) was manually recreated it is then given a user OID and therefore no longer "seen" as a system database so not automatically included in the restore. Currently, if this has happened, you could specify --db-include=postgres since it no longer has a system OID but with this PR, we will automatically include it again.

**Comment by woblerr** [comment]
@cmwshang  Thank you, I understand. Will further implementation of the flag 'db-excluse' be carried out within a separate PR?

**Comment by cmwshang** [comment]
> @cmwshang Thank you, I understand. Will further implementation of the flag 'db-excluse' be carried out within a separate PR?

A possible 'db-exclude' option is being considered - but generally, any database that you do not specifically list with --db-include will be excluded.

**Comment by pgstef** [comment]
> After further discussions, we decided we should split this PR and get the original issue solved first. The db-exclude option adds a lot of code and tests for something that we feel we need to think more about. So, @pgstef if you could pull out the db-exclude code and leave in the code and tests for getting the systemIds from the names, that would be great.

I can't say I'm not a little bit disappointed by this answer, since I followed Stephen's intuition regarding over-sized postgres databases. I strongly believe the exclude option would be great/helpful.

However, I removed the exclude part and force-pushed after refreshing the branch upon master. Imho, the exclude option doesn't really add so much code.

I'm still waiting that the MockAllTest completes but I think it will break regarding we rely on databases names/ids in the manifest to detect the systemIds. So we'll need to fix those tests to add those names to the manifest but I still didn't figured how to do that unfortunately :-(

**Comment by pgstef** [comment]
@cmwshang, I've added some safe-guard in commit cfbf43c to allow the MockAllTest to succeed even if I don't really like that solution. If you find something prettier in MockAllTest, I think this specific commit could be reverted ;-)

**Comment by dwsteele** [comment]
@pgstef We are still interested in the exclude option, but feel it was too limited as specified. Having an entire option and tests to for the purpose of excluding a single database that is almost never large seemed too much. We would prefer the exclude option to be general purpose, for instance if I have 100 dbs but only 2 large ones, it would be nice to be able to restore all the small ones by excluding the two large ones. By limiting exclude to system databases it ends up being a lot less useful.

So, as we often do, we want to split this PR to fix the original reported issue (postgres database was not restored automatically) and then consider exclude separately.

**Comment by pgstef** [comment]
> @pgstef We are still interested in the exclude option, but feel it was too limited as specified. Having an entire option and tests to for the purpose of excluding a single database that is almost never large seemed too much. We would prefer the exclude option to be general purpose, for instance if I have 100 dbs but only 2 large ones, it would be nice to be able to restore all the small ones by excluding the two large ones. By limiting exclude to system databases it ends up being a lot less useful.

Thanks for the clarification David. The original code removed today allowed to exclude any db, not only system ones. But I see your point to split it into 2 PR's. I'll submit the code again when this one will be completed :-)

**Comment by cmwshang** [comment]
> Thanks for the clarification David. The original code removed today allowed to exclude any db, not only system ones. But I see your point to split it into 2 PR's. I'll submit the code again when this one will be completed :-)

Yes, the original code could be used to specify any database, but it does not make any sense since anything not specified by --db-include, is **automatically** excluded except system databases, so say you have 10 database mydb1 through mydb10 what would it mean to say --db-include=mydb1 --db-exclude=mydb2 - the exclude is a NOOP - it makes no sense because --db-include=mydb1 will exclude all other user databases (mydb2 through mydb10) anyway. If the --db-exclude=postgres, then it makes sense because we would normally have included postgres but now we are being told not to.  Since we would like to fix the original bug identified in the issue and get it into the next release, which is happening soon, then splitting this PR to allow for the bug to be fixed, and then address an exclude feature, allows us more time to consider how best to implement it that would be the most beneficial.

**Comment by pgstef** [comment]
> Yes, the original code could be used to specify any database, but it does not make any sense since anything not specified by --db-include, is **automatically** excluded except system databases, so say you have 10 databases mydb1 through mydb10 what would it mean to say --db-include=mydb1 --db-exclude=mydb2 - the exclude is a NOOP - it makes no sense because --db-include=mydb1 will exclude all other user databases (mydb2 through mydb10) anyway. 

Understood. That's indeed the complexity of combining include and exclude. Which one will take precedence,... In my mind, `include=mydb1 exclude=mydb2` will result in excluding everything, so mydb2 will indeed be excluded. That's what I would expect if I combine both options.

But you're right, let's focus on this _include system-dbs_ first and then I'll submit a new PR where we can discuss this :-)

**Comment by cmwshang** [comment]
> But you're right, let's focus on this _include system-dbs_ first and then I'll submit a new PR where we can discuss this :-)

Great! Let's circle back once we complete this PR and come up with some requirements for the exclude feature.

File: src/command/restore/restore.c [hunk 11]
```diff
@@ -1320,6 +1320,21 @@ restoreSelectiveExpression(Manifest *manifest)
             // Log databases found
             LOG_DETAIL_FMT("databases found for selective restore (%s)", strZ(strLstJoin(dbList, ", ")));
 
+            // Get all system db ids, even in case users would have manually re-created it
```
Comment (cmwshang) [inline src/command/restore/restore.c:1323]: I made some changes and I hope you will agree with them. I realized a few things in doing live tests:

1) Unlike the unit tests, the system databases are displayed in the `databases found for selective restore (%s)` message. So the unit tests were a bit lacking here because the PG_VERSION is not always added in the manifest in the unit tests when, realistically, they should be. But since you fixed the issue, we'd always display the system ids, as we should, in the `databases found for selective restore (%s)` message. I only updated one unit test as an example, because I wanted to give you an opportunity to review the changes I made and incorporate them into your branch if you agree and maybe update the unit tests yourself.

2) We have a cool anti-merge function (strLstMergeAnti) that was useful here so, using that, we are able to use the dbList array as the main array for verifying and checking databases so I moved the generation of the systemDbIdList up and added the ids to the dbList (rather than remove). And also was able to remove the need to keep checking the PG_USER_OBJECT_MIN_ID.

3) We have gotten complaints on the amount of logging we do, so, given the strLstMergeAnti() function we're able filter the dbList with it and then provide the list of zeroed databases in a concatenated string, just like we do with the `databases found for selective restore (%s)` message.

I tested before and after the change on a live system and everything is working as expected. I have not updated the unit tests or the mock tests. I can definitely take care of the mock tests, but, again, I wanted to see what you thought of the changes and see if you wanted to try to merge them into your branch and work on the unit tests. 

The comparison of my branch to your branch is at:
https://github.com/pgstef/pgbackrest/compare/issues/986...cmwshang:pgstef-bugfix-only-issue986

And my branch itself is at: https://github.com/cmwshang/pgbackrest/tree/pgstef-bugfix-only-issue986)

Please let me know your thoughts.

**Review by cmwshang — COMMENTED** [review]
See comment.

File: src/command/restore/restore.c (hunk 11 above)
Comment (pgstef) [reply src/command/restore/restore.c]: I like your changes **a lot**, thanks @cmwshang!
I struggled to merge your branch but I managed to extract and import your patch in this PR. I hope that's not a problem :-(
I fixed the unit tests too. I'm now waiting to see if the MockAllTest will succeed.

File: src/command/restore/restore.c (hunk 10 above)
[the remark this answers is not in the document]
Comment (pgstef) [reply src/command/restore/restore.c]: Seems the code cover this exception now:

```
if (cvtZToUInt64(strZ(dbId)) < PG_USER_OBJECT_MIN_ID)
                   strLstAddIfMissing(systemDbIdList, dbId);
```

**Review by cmwshang — APPROVED** [review]
Review complete and approved. Awesome job, Stefan!

**Comment by pgstef** [comment]
I don't think I got write permissions on the repo but perhaps next time we'd like to collaborate, it would be handy to have a common "dev-.." branch here rather in our forks?

> Review complete and approved. 

Thanks a lot for your review and help!

**Comment by dwsteele** [comment]
> I don't think I got write permissions on the repo but perhaps next time we'd like to collaborate, it would be handy to have a common "dev-.." branch here rather in our forks?

I can get Cynthia permission to modify PRs like I do -- we just need to get around to setting it up.

**Comment by dwsteele** [comment]
@pgstef If you are interested in pursuing the patch to add `--db-exclude`, I think that the way it was defined originally is too narrow to make the feature worth it.

I'd like to see `--db-exclude` be usable by itself, so for example I could restore all databases except one or two especially large ones. When used with `--db-include` it would only affect system databases, though it probably wouldn't make sense to error unless there is an intersection between `--db-include` and `--db-exclude`.
